keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -51,223 +48,183 @@
-        * And if the in-tree volume plugin supports SELinux / `CSIDriver.Spec.SELinuxMountSupported` is explicitly `true`:
-            * Kubelet tries to mount the volume for the Pod with given SELinux label using `mount -o context=XYZ`.
-                * Kubelet makes sure the option is passed to the first mount in all in-tree volume plugins (incl. ephemeral volumes like Secrets).
-                * Kubelet passes it as a mount option to all CSI calls for given volume.
-            * After the volume is mounted, kubelet checks that the root of the volume has the expected SELinux label, i.e. that the volume was mounted correctly.
-                * If the volume root has expected label, kubelet passes the volume to the container runtime without any ":z" or ":Z" options - no relabeling is necessary.
-                * If the volume root has unexpected label, for example when CSI driver did not apply `-o context` correctly, or the volume was already mounted with a different context,
-                  volume plugin reports an error and kubelet fails to start the pod.
-                  It is CSI driver fault that it advertises SELinux support and then fails to apply it.
-
-* Nothing changes when `CSIDriver.Spec.SELinuxMountSupported` is `false` or not set:
-    * CSI volume plugin calls CSI without any special SELinux mount options and it autodetects, if the volume supports SELinux or not by presence of `seclabel` mount option.
-      This is current kubelet behavior.
-
-* Nothing changes if kubelet does not know the SELinux context of a pod (`SELinuxOptions.Level` is empty) or pod's `SELinuxRelabelPolicy` is `Always`.
-    * Volume is mounted without any SELinux options and passed to the container runtime with or without ":Z", depending on if the volume plugin supports SELinux or not (by checking `seclabel` mount option).
-      The container runtime allocates a new SELinux context and recursively relabels all files on the volume.
-      This is current kubelet behavior.
-
-Validation:
+#### Behavioral changes
+This KEP changes behavior of Kubernetes when two pods with different SELinux contexts use the same volume.
+Let Pod A with SELinux context X runs and Pod B with SELinux context Y is about to start on the same node and both use the same volume.
 
-* Kubernetes checks that `SELinuxRelabelPolicy` field can be used in a pod only when at least `SELinuxOptions.Level` is set.
+* *Before this KEP*: Pod A suddenly starts getting "permission denied" errors when accessing files on the volume, because the container runtime re-labeled all files on it with label Y when starting pod B. Pod B will start just fine and can access the volume.
+* *As proposed in this KEP*: Pod B won't even start, because the volume is already mounted with `-o context=X`. When kubelet tries to mount the same volume with `-o context=Y`, this mount fails. The Pod B with be `ContainerCreating` until Pod A is deleted and its volumes unmounted.
```

Comment:
Will we have events/messages indicating why Pod B stays in ContainerCreating?

Reply:
The message is a generic `mount: wrong fs type, bad option, bad superblock on /dev/sdb, missing codepage or helper program, or other error`. `/bin/mount` does not tell which mount option is the bad one.

Added a note about it.

Reply:
> The message is a generic mount: wrong fs type, bad option, bad superblock on /dev/sdb, missing codepage or helper program, or other error. /bin/mount does not tell which mount option is the bad one.

Can external logic on a node look at the mount that is failing and determine if it is being mounted twice with different `-o context=foo` options?  That message is really hard for laymen to interpret and doesn't give much hint to the pod author.

Reply:
I added some wording around - kubelet will track its mounts and with what context they were set up, so it may throw a nice error like `volume X is already used by pod Y with another SELinux context`.